### DISCUSSION TITLE
Setting the current theme via de D-Bus API

### DISCUSSION THREAD

**Discussion by WhyNotHugo** [audacious-media-player/audacious#114] [General]
I'm trying to programatically change the theme. I want it to change to light mode at sunrise and dark mode and sundown. For this I am using `darkman`, which can just run a script of my choosing at the right time.

To change the theme, I'm calling the D-Bus API exposed by audacious. When I change the theme, the setting gets updated, but the theme remains the same. I will elaborate further.

When running the following command:

    gdbus call --session --dest org.atheme.audacious --object-path /org/atheme/audacious --method org.atheme.audacious.ConfigGet audqt theme

If the current theme is the default, it prints `''`.
If the current theme is the dark one, it prints `'dark'`

This seems to confirm that I'm reading the right setting.

I can edit the setting with any of the following two commands:

    gdbus call --session --dest org.atheme.audacious --object-path /org/atheme/audacious --method org.atheme.audacious.ConfigSet audqt theme ''
    gdbus call --session --dest org.atheme.audacious --object-path /org/atheme/audacious --method org.atheme.audacious.ConfigSet audqt theme 'dark'

When I edit the setting, and the open settings in audacious, the new value is selected in the combo. But the theme **does not** change. If I try to `ConfigGet` the theme setting, the updated value is also returned, but the theme that audacious is actually using does not match the value.

I think that when I change the theme, audacious updates it's internal state of settings, but it doesn't apply the theme itself. Do I have to do something different, or have I stumbled across a bug?

**Comment by radioactiveman** [comment]
Some settings may not be applied immediately and require a restart of Audacious. Could you not do that in your script as some kind of workaround?

Audacious itself calls the method below when changing the theme in the settings dialog, but probably not when using the D-Bus approach.

https://github.com/audacious-media-player/audacious/blob/a9d628098f917f42c6971e36392458e50f0cc9ec/src/libaudqt/prefs-window-qt.cc#L240-L246

https://github.com/audacious-media-player/audacious/blob/e2f07fb4c7da499de819476d67c193e05d02148a/src/libaudqt/dark-theme.cc#L136-L149

**Comment by WhyNotHugo** [comment]
The dbus call seems to go through `src/audacious/dbus-server.cc`, function `do_config_set`. Indeed, it simply updates the status table and nothing else (I guess maybe some other config settings don't apply immediately either?).

I guess that I can add a check there and `if section ==  "audqt" && name =="theme"` then call `theme_changed`. Is `strcmp` okay to use in this context?

**Comment by WhyNotHugo** [comment]
> require a restart of Audacious. Could you not do that in your script as some kind of workaround?

This script triggers a couple of times a day automatically (e.g.: I don't run it manually). It would be kinda awkward since the new instance would interrupt whatever I'm doing at the time.
